File: src/integration/jackson/src/jvmMain/java/community/flock/wirespec/integration/jackson/java/WirespecSerialization.java
package community.flock.wirespec.integration.jackson.java;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import community.flock.wirespec.java.Wirespec.Serialization;
import community.flock.wirespec.java.serde.DefaultParamSerialization;
import community.flock.wirespec.java.serde.DefaultPathSerialization;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;

/**
 * A reusable implementation of Wirespec.Serialization that uses Jackson for serialization and deserialization.
 * This class implements parameter serialization and deserialization using a private ParamSerialization field.
 */
public class WirespecSerialization implements Serialization, DefaultParamSerialization, DefaultPathSerialization {

    private final ObjectMapper wirespecObjectMapper;

    public WirespecSerialization(ObjectMapper objectMapper) {
        this.wirespecObjectMapper = objectMapper.copy()
                .registerModule(new WirespecModuleJava());
    }

    @Override
    public <T> byte[] serializeBody(T body, Type type) {
        if (body instanceof String) {
            return ((String) body).getBytes(StandardCharsets.UTF_8);
        } else {
            try {
                return wirespecObjectMapper.writeValueAsBytes(body);
            } catch (IOException e) {
                throw new RuntimeException("Failed to serialize body", e);
            }
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T> T deserializeBody(byte[] raw, Type valueType) {
        if (raw == null) {
            return null;
        }

        if (valueType.equals(String.class)) {
                return (T) new String(raw, StandardCharsets.UTF_8);
            } else {
            try {
                JavaType type = wirespecObjectMapper.getTypeFactory().constructType(valueType);
                return wirespecObjectMapper.readValue(raw, type);
            } catch (IOException e) {
                throw new RuntimeException("Failed to deserialize body", e);
            }
        }
    }
}

File: src/integration/jackson/src/jvmMain/kotlin/community/flock/wirespec/integration/jackson/kotlin/WirespecSerialization.kt
package community.flock.wirespec.integration.jackson.kotlin

import com.fasterxml.jackson.databind.ObjectMapper
import community.flock.wirespec.kotlin.Wirespec
import community.flock.wirespec.kotlin.Wirespec.Serialization
import community.flock.wirespec.kotlin.serde.DefaultParamSerialization
import community.flock.wirespec.kotlin.serde.DefaultPathSerialization
import kotlin.reflect.KType
import kotlin.reflect.javaType

/**
 * A reusable implementation of Wirespec.Serialization that uses Jackson for serialization and deserialization.
 * This class implements parameter serialization and deserialization using a private ParamSerialization field.
 */
class WirespecSerialization(
    objectMapper: ObjectMapper,
) : Serialization,
    Wirespec.ParamSerialization by DefaultParamSerialization(),
    Wirespec.PathSerialization by DefaultPathSerialization() {

    private val wirespecObjectMapper = objectMapper.copy().registerModule(WirespecModuleKotlin())

    override fun <T> serializeBody(t: T, kType: KType): ByteArray = when (t) {
        is String -> t.toByteArray()
        else -> wirespecObjectMapper.writeValueAsBytes(t)
    }

    @Suppress("UNCHECKED_CAST")
    @OptIn(ExperimentalStdlibApi::class)
    override fun <T> deserializeBody(raw: ByteArray, kType: KType): T = when {
        kType.classifier == String::class -> raw as T
        else ->
            wirespecObjectMapper
                .constructType(kType.javaType)
                .let { wirespecObjectMapper.readValue(raw, it) }
    }
}
